platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/HiDPIUtils.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.swing.laf.flatlaf;
+
+import java.awt.Graphics;
+import java.awt.Graphics2D;
+import java.awt.geom.AffineTransform;
+
+public class HiDPIUtils {
+
+    public interface HiDPIPainter {
+        void paint(Graphics2D g, int width, int height, double scale);
+    }
+
+    /**
+     * Paint at scale factor 1x to avoid rounding issues at 125%, 150% and 175% scaling.
+     * <p>
+     * Scales the given Graphics2D down to 100% and invokes the
+     * given painter passing scaled deviceWidth and deviceHeight.
+     * <p>
+     * Transformation to device pixels borrowed from class
+     * {@link org.netbeans.swing.plaf.windows8.DPISafeBorder}.
+     */
+    public static void paintAtScale1x(Graphics g0, int x, int y, int width, int height, HiDPIPainter painter) {
```

Comment:
Again, it might be nice to factor out this code eventually, but we can wait until we have seen a few different uses of it.

Reply:
👍  

I've just avoided to add new API to other modules...